### PR TITLE
Use separate variables for "expr != null" conditions. This allows to reu...

### PR DESCRIPTION
**Pull request by vlsi**
...se the result of expr calculation

This is a basic reuse of expressions that might duplicate out of "null-checks".
For instance: `upper(udf(1))` is typically translated to `udf(1) == null ? null : upper(udf(1))`
The fix parttern-matches `XXX == null` and factors `XXX` to separate declaration.
Thus final expression becomes `v0 = udf(1); v1 = v0 == null ? null : upper(v0)`

The fix requires equals/hashCode support from linq4j: https://github.com/julianhyde/linq4j/pull/12

### CODE REVIEW

**Comment by julianhyde** [comment]
Can you please provide a test case? Does this pull request depend on any of your other open issues: #236, #228, #231, #225? I'd like to check-in the issues that don't require a linq4j change first.

**Comment by vlsi** [comment]
This is the only PR that requires linq4j update.

Ok, will add some tests

**Comment by vlsi** [comment]
Proper equals/hashCode breaks behaviour when "optiq starts to evaluate arguments before null checks".
Even without #235, with just upgraded linq4j, `ReflectiveSchemaTest.testDivide`:

``` sql
select "wrapperLong" / "wrapperLong" as c from "s"."everyTypes" where "primitiveLong" <> 0
```

``` java
public Object current() {
  final Long v = ((net.hydromatic.optiq.test.ReflectiveSchemaTest.EveryType) inputEnumerator.current()).wrapperLong;
  final long v1 = v.longValue();
  return v == null ? (Long) null : Long.valueOf(v1 / v1);
}
```

Note how optiq keeps `v.longValue()` in a local variable.
This will throw `NPE` before it wil check `v==null` :(

v1 appears when performing `translator.translateList(call.getOperands())` under the assumption `v != null`

**Comment by vlsi** [comment]
Rebased against updated linq4j: `JavaRules.implement(EnumerableRelImplementor` was too optimistic when passing `optimize=true` for mutable expressions (e.g. `new SortedMap`).

`ReflectiveSchemaTest.testDivide` fails as listed above.

**Comment by vlsi** [comment]
What do you think if translate `return expr == null ? null : fun(expr)` as
`int tmp; return (tmp=expr) == null ? null : fun(tmp)` ?

In other words:
1) create non-final locals for reuse of expr
2) when translating 'null not possible' part, forbid creating of new locals.

Another way to solve 'arguments are calculated too early' issue is to use
`if (args are null) { res = null; } else { res = fun(expr); }` kind of
expressions.

The big problem is we'll need to double-check existing expression generator
code if it contains similar cases when parts of guarded expression are
evaluated earlier than the guard itself.

**Comment by julianhyde** [comment]
I like that idea.

**Comment by julianhyde** [comment]
@vlsi I get 1 failure, 1 error running against resuse_exprs_for_null_implemetor_conditions branch plus the latest linq4j as of https://github.com/julianhyde/linq4j/commit/a9f47857764254acc08d24786e0278c57b3bf9f1.

```
Tests run: 19, Failures: 1, Errors: 1, Skipped: 3, Time elapsed: 14.324 sec <<< FAILURE! - in net.hydromatic.optiq.test.ReflectiveSchemaTest
testDivide(net.hydromatic.optiq.test.ReflectiveSchemaTest)  Time elapsed: 0.061 sec  <<< FAILURE!
testCompareJavaAndSqlTypes(net.hydromatic.optiq.test.ReflectiveSchemaTest)  Time elapsed: 0.062 sec  <<< ERROR!
```

Can you fix these please?

**Comment by vlsi** [comment]
Julian, that is exact issue that I mention earlier: arguments are evaluated
too early since translate arguments blindly uses 'optimize=true', thus it
results in a local variable before ternary.

I believe this was not seen earlier since optiq just could not find
matching expressions since equals was not there.

As a quick fix we can switch to 'optimize=false' when transtating
arguments-null-check _and_ translateCall itself.

I am not sure how long it will take me to refactor call translation to if
(){} else {} strategy.

**Comment by julianhyde** [comment]
What is the effect of your proposed 'quick fix'? Does it impact performance? I don't want to adopt a quick fix if we don't have a plan to do it properly.

Is there a simpler change we can make to optiq so that it can use the latest linq4j. (I hope that the current linq4j HEAD will become linq4j-0.2 in the next day or two.)

It doesn't seem likely that this pull request will be fully resolved in the next few days. Do you agree? If so, I propose that you log an issue to track the problem. That will take the pressure off so that we can think how we really want to solve this.

**Comment by vlsi** [comment]
I expect no performance degradation if we use optimize=false.

At least with old linq4j this optimize=true never worked.

**Comment by vlsi** [comment]
I've attempted `if () {}` strategy.

Here are the changes:
https://github.com/vlsi/linq4j/commit/169efe906dea2e07b328b20f56d98966a4e292be
https://github.com/vlsi/optiq/commit/27a731273372166711b3597c78301c657a0623fd

Basically I add nesting API for `BlockBuilder` and `Translator` and hand-write the nested ifs.

The bad thing with `if () { }` approach is it is hard to reuse computed expressions.
for instance:

``` sql
select
 "wrapperLong" / "wrapperLong" as c 
,"wrapperLong" / "wrapperLong" as d 
,"wrapperLong" / "wrapperLong" + "wrapperLong" / "wrapperLong" as c
 from "s"."everyTypes" where "primitiveLong" <> 0
```

Rex does reuse computations across different projections. Maybe the way to go is to cache `<Pair<RexNode, NullsAs>, DeclarationStatement>`, however it looks ugly.

Current prototype generates the following code:

``` java
public Object current() {
  final Long v = ((net.hydromatic.optiq.test.ReflectiveSchemaTest.EveryType) inputEnumerator.current()).wrapperLong;
  final boolean v0 = v != null;
  Long res = (Long) null;
  if (v0) {
    final long v1 = v.longValue();
    res = Long.valueOf(v1 / v1);
  }
  Long res0 = (Long) null;
  if (v0) {
    final long v1 = v.longValue();
    res0 = Long.valueOf(v1 / v1); // this is better be taken from res directly
  }
  Long res1 = (Long) null;
  if (v0 && v0) {
    final long v2 = v / v;
    res1 = Long.valueOf(v2 + v2);
  }
  return new Object[] {
      res,
      res0,
      res1};
```

Basically, the results of division `v/v` are unshared across projections (`res`, `res0`, `res1`) since `ParameterExpression` has identity equals-hashcode semantics, thus `v1` for the `res` is not equal to the `v1` for the `res0`.

**Comment by julianhyde** [comment]
I assume that the issue with optimize=true could be demonstrated in linq4j. If so, I suggest a compromise. I will set optimize=false and accept this patch into optiq-0.6 and also release linq4j-0.2 based upon https://github.com/julianhyde/linq4j/commit/a9f47857764254acc08d24786e0278c57b3bf9f1 provided that you log an issue against linq4j with a test case that shows BlockBuilder generating invalid code. After linq4j-0.2 you & I can work together to fix it and re-enable optimize=true.

**Comment by vlsi** [comment]
This generally works for me.
Do you mind fixing `Expressions.ifThenElse`? (i.e. `ConditionalStatement#accept0(ExpressionWriter writer)` is flawed)

> I will set optimize=false and accept this patch

To be correct: not this patch, but just corrections to optiq so it uses `optimize=false` properly (`JavaRules` part of the patch since current implementation of Window aggregates is too optimistic to share `SortedMultiMap`).

`expression instanceof BinaryExpression` kind of changes are not safe to include yet.

**Comment by julianhyde** [comment]
I've fixed https://github.com/julianhyde/linq4j/issues/15 (ifThenElse).

The ball is now in your court. Create (1) a pull request on optiq that passes all tests on the latest linq4j, and (2) a pull request on linq4j that contains a disabled test case for the code-generation issues. If you do this today I will include linq4j-0.2 in the optiq-0.6 release. If not, we'll get to it after the release.

**Comment by vlsi** [comment]
Please find a temporary fix for optiq vs linq4j compatibility in https://github.com/julianhyde/optiq/pull/248.

I am not ready to push `if` version yet.

**Comment by vlsi** [comment]
I've created a couple of pull requests to linq4j (https://github.com/julianhyde/linq4j/pull/16) and optiq(https://github.com/julianhyde/optiq/pull/249).

Those commits eliminate boxing in several cases.
There is still room for improvement (i.e. `CaseImplementor` does not use `nullAs == NOT_POSSIBLE` hint, thus it generates always boxed code)

**Comment by julianhyde** [comment]
I'll accept the linq4j pull request.

I can't accept the optiq pull request until there is a new release of linq4j. Please advise on timing on that release. But will be at least a couple of weeks.

**Comment by vlsi** [comment]
I do not have a specific date in mind.

I just try to split the work to keep it and/or the review manageable.

**Comment by julianhyde** [comment]
Fixed in https://github.com/julianhyde/optiq/commit/cff9fb03703caff1f74e2c6c4d410d8a64fb8667. The code in that commit has evolved since this original pull request, but the tests are included and all pass, so I declare it done.
